File: sdk/rust/src/artifact/step.rs
use crate::{
    api,
    api::artifact::ArtifactSystem::{self, Aarch64Darwin, Aarch64Linux, X8664Darwin, X8664Linux},
    artifact::{self, linux_vorpal::LinuxVorpal},
    context::ConfigContext,
};
use anyhow::{bail, Result};
use indoc::formatdoc;

/// Builds a plain `bash` step that runs `script` with the given artifacts on
/// `PATH` and the given extra environment variables set.
#[must_use]
pub fn bash(
    artifacts: &[String],
    environments: &[String],
    secrets: &[api::artifact::ArtifactStepSecret],
    script: &str,
) -> api::artifact::ArtifactStep {
    let mut step_environments: Vec<String> = environments
        .iter()
        .filter(|environment| !environment.starts_with("PATH="))
        .cloned()
        .collect();

    let step_path_bins = artifacts
        .iter()
        .map(|a| format!("{}/bin", artifact::get_env_key(a)))
        .collect::<Vec<String>>()
        .join(":");

    let step_path_default = "/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin";

    let mut step_path = format!("{step_path_bins}:{step_path_default}");

    if let Some(path) = environments.iter().find(|x| x.starts_with("PATH=")) {
        if let Some(path_value) = path.split('=').nth(1) {
            step_path = format!("{path_value}:{step_path}");
        }
    }

    step_environments.push("HOME=$VORPAL_WORKSPACE".to_string());
    step_environments.push(format!("PATH={step_path}"));

    let step_script = formatdoc! {"
        #!/bin/bash
        set -euo pipefail

        {script}
    "};

    artifact::ArtifactStep::new("bash")
        .with_artifacts(artifacts.to_vec())
        .with_environments(step_environments)
        .with_secrets(secrets.to_vec())
        .with_script(step_script)
        .build()
}

/// Builds a `bwrap` (bubblewrap) sandboxed step that runs `script` inside a
/// minimal, isolated root filesystem assembled from `rootfs` and `artifacts`.
///
/// # Errors
///
/// Currently infallible; the `Result` is retained so callers can treat
/// `bwrap` and [`shell`] uniformly.
#[expect(
    clippy::unused_async,
    reason = "public SDK API called as `.await` from every artifact module; removing async is a breaking change"
)]
#[expect(
    clippy::too_many_lines,
    reason = "assembles one flat bwrap argument list; splitting it would scatter one linear construction across helpers without adding clarity"
)]
pub async fn bwrap(
    arguments: &[&str],
    artifacts: &[String],
    environments: &[String],
    rootfs: Option<&str>,
    secrets: &[api::artifact::ArtifactStepSecret],
    script: String,
) -> Result<api::artifact::ArtifactStep> {
    // Setup arguments

    let mut step_arguments = vec![
        "--unshare-all".to_string(),
        "--share-net".to_string(),
        "--clearenv".to_string(),
        "--chdir".to_string(),
        "$VORPAL_WORKSPACE".to_string(),
        "--gid".to_string(),
        "1000".to_string(),
        "--uid".to_string(),
        "1000".to_string(),
        "--dev".to_string(),
        "/dev".to_string(),
        "--proc".to_string(),
        "/proc".to_string(),
        "--tmpfs".to_string(),
        "/tmp".to_string(),
        "--bind".to_string(),
        "$VORPAL_OUTPUT".to_string(),
        "$VORPAL_OUTPUT".to_string(),
        "--bind".to_string(),
        "$VORPAL_WORKSPACE".to_string(),
        "$VORPAL_WORKSPACE".to_string(),
        "--setenv".to_string(),
        "VORPAL_OUTPUT".to_string(),
        "$VORPAL_OUTPUT".to_string(),
        "--setenv".to_string(),
        "VORPAL_WORKSPACE".to_string(),
        "$VORPAL_WORKSPACE".to_string(),
        "--setenv".to_string(),
        "HOME".to_string(),
        "$VORPAL_WORKSPACE".to_string(),
    ];

    // Setup artifacts

    let mut step_artifacts: Vec<String> = vec![];

    if let Some(rootfs) = rootfs {
        let rootfs_env = artifact::get_env_key(rootfs);

        let rootfs_args = vec![
            "--ro-bind".to_string(),
            format!("{rootfs_env}/bin"),
            "/bin".to_string(),
            "--ro-bind".to_string(),
            format!("{rootfs_env}/etc"),
            "/etc".to_string(),
            "--ro-bind".to_string(),
            format!("{rootfs_env}/lib"),
            "/lib".to_string(),
            "--ro-bind-try".to_string(),
            format!("{rootfs_env}/lib64"),
            "/lib64".to_string(),
            "--ro-bind".to_string(),
            format!("{rootfs_env}/sbin"),
            "/sbin".to_string(),
            "--ro-bind".to_string(),
            format!("{rootfs_env}/usr"),
            "/usr".to_string(),
        ];

        step_arguments.extend(rootfs_args);
        step_artifacts.push(rootfs.to_string());
    }

    // Setup artifact arguments

    step_artifacts.extend_from_slice(artifacts);

    for artifact in &step_artifacts {
        step_arguments.push("--ro-bind".to_string());
        step_arguments.push(artifact::get_env_key(artifact));
        step_arguments.push(artifact::get_env_key(artifact));
        step_arguments.push("--setenv".to_string());
        step_arguments.push(artifact::get_env_key(artifact).replace('$', ""));
        step_arguments.push(artifact::get_env_key(artifact));
    }

    // Setup environment arguments

    let step_path_bins = step_artifacts
        .iter()
        .map(|a| format!("{}/bin", artifact::get_env_key(a)))
        .collect::<Vec<String>>()
        .join(":");

    let mut step_path = format!("{step_path_bins}:/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin");

    if let Some(path) = environments.iter().find(|x| x.starts_with("PATH=")) {
        if let Some(path_value) = path.split('=').nth(1) {
            step_path = format!("{path_value}:{step_path}");
        }
    }

    step_arguments.push("--setenv".to_string());
    step_arguments.push("PATH".to_string());
    step_arguments.push(step_path);

    for env in environments {
        let Some((key, value)) = env.split_once('=') else {
            continue;
        };

        if key == "PATH" {
            continue;
        }

        step_arguments.push("--setenv".to_string());
        step_arguments.push(key.to_string());
        step_arguments.push(value.to_string());
    }

    // Setup arguments

    for argument in arguments {
        step_arguments.push((*argument).to_string());
    }

    // Setup script

    let step_script = formatdoc! {"
        #!/bin/bash
        set -euo pipefail

        {script}
    "};

    // Setup step

    let step = artifact::ArtifactStep::new("bwrap")
        .with_arguments(
            step_arguments
                .iter()
                .map(std::string::String::as_str)
                .collect(),
        )
        .with_artifacts(step_artifacts)
        .with_environments(vec![
            "PATH=/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin".to_string()
        ])
        .with_secrets(secrets.to_vec())
        .with_script(step_script)
        .build();

    Ok(step)
}

/// Builds a step appropriate for the context's target system: a plain `bash`
/// step on Darwin, or a `bwrap`-sandboxed step backed by `linux-vorpal` on
/// Linux.
///
/// # Errors
///
/// Returns an error if the context's target system is [`ArtifactSystem::UnknownSystem`],
/// or if building the `linux-vorpal` root filesystem for a Linux target fails.
pub async fn shell(
    context: &mut ConfigContext,
    artifacts: &[String],
    environments: &[String],
    script: String,
    secrets: &[api::artifact::ArtifactStepSecret],
) -> Result<api::artifact::ArtifactStep> {
    // Setup target

    let step_system = context.get_system();

    // Setup step

    let step = match step_system {
        Aarch64Darwin | X8664Darwin => bash(artifacts, environments, secrets, &script),

        Aarch64Linux | X8664Linux => {
            let linux_vorpal = LinuxVorpal::new().build(context).await?;

            bwrap(
                &[],
                artifacts,
                environments,
                Some(linux_vorpal.as_str()),
                secrets,
                script,
            )
            .await?
        }

        ArtifactSystem::UnknownSystem => {
            bail!("unsupported system: {}", step_system.as_str_name())
        }
    };

    Ok(step)
}

// TODO: Add support for secrets with docker step

/// Builds a `docker` step that runs the given `docker` CLI arguments against
/// the given artifacts.
#[must_use]
pub fn docker(arguments: &[&str], artifacts: &[String]) -> api::artifact::ArtifactStep {
    artifact::ArtifactStep::new("docker")
        .with_arguments(arguments.to_vec())
        .with_artifacts(artifacts.to_vec())
        .with_environments(vec![
            "PATH=/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin".to_string()
        ])
        .build()
}

File: sdk/typescript/src/artifact/step.ts
import type {
  ArtifactStep as ArtifactStepMsg,
  ArtifactStepSecret,
} from "../api/artifact/artifact.js";
import { ArtifactSystem } from "../api/artifact/artifact.js";
import type { ConfigContext } from "../context.js";
import { getEnvKey } from "../artifact.js";

/**
 * Creates a bash step. Matches Rust sdk/rust/src/artifact/step.rs bash().
 *
 * - Filters PATH from environments
 * - Builds PATH from artifact bins + default system paths
 * - Prepends bash shebang and set -euo pipefail to script
 * - Uses "bash" as entrypoint
 */
export function bash(
  artifacts: string[],
  environments: string[],
  secrets: ArtifactStepSecret[],
  script: string,
): ArtifactStepMsg {
  const stepEnvironments: string[] = [];

  for (const environment of environments) {
    if (environment.startsWith("PATH=")) {
      continue;
    }
    stepEnvironments.push(environment);
  }

  const stepPathBins = artifacts
    .map((a) => `${getEnvKey(a)}/bin`)
    .join(":");

  const stepPathDefault = "/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin";

  let stepPath = `${stepPathBins}:${stepPathDefault}`;

  const pathEnv = environments.find((x) => x.startsWith("PATH="));
  if (pathEnv) {
    const pathValue = pathEnv.split("=").slice(1).join("=");
    if (pathValue) {
      stepPath = `${pathValue}:${stepPath}`;
    }
  }

  stepEnvironments.push("HOME=$VORPAL_WORKSPACE");
  stepEnvironments.push(`PATH=${stepPath}`);

  const stepScript = `#!/bin/bash\nset -euo pipefail\n\n${script}\n`;

  // Deduplicate secrets by name
  const uniqueSecrets: ArtifactStepSecret[] = [];
  for (const secret of secrets) {
    if (!uniqueSecrets.some((s) => s.name === secret.name)) {
      uniqueSecrets.push(secret);
    }
  }

  return {
    entrypoint: "bash",
    script: stepScript,
    secrets: uniqueSecrets,
    arguments: [],
    artifacts: artifacts,
    environments: stepEnvironments,
  };
}

/**
 * Creates a bwrap step. Matches Rust sdk/rust/src/artifact/step.rs bwrap().
 *
 * CRITICAL: Argument list ordering must be identical to Rust/Go.
 */
export function bwrap(
  arguments_: string[],
  artifacts: string[],
  environments: string[],
  rootfs: string | null,
  secrets: ArtifactStepSecret[],
  script: string,
): ArtifactStepMsg {
  // Setup arguments
  const stepArguments: string[] = [
    "--unshare-all",
    "--share-net",
    "--clearenv",
    "--chdir",
    "$VORPAL_WORKSPACE",
    "--gid",
    "1000",
    "--uid",
    "1000",
    "--dev",
    "/dev",
    "--proc",
    "/proc",
    "--tmpfs",
    "/tmp",
    "--bind",
    "$VORPAL_OUTPUT",
    "$VORPAL_OUTPUT",
    "--bind",
    "$VORPAL_WORKSPACE",
    "$VORPAL_WORKSPACE",
    "--setenv",
    "VORPAL_OUTPUT",
    "$VORPAL_OUTPUT",
    "--setenv",
    "VORPAL_WORKSPACE",
    "$VORPAL_WORKSPACE",
    "--setenv",
    "HOME",
    "$VORPAL_WORKSPACE",
  ];

  // Setup artifacts
  const stepArtifacts: string[] = [];

  if (rootfs !== null) {
    const rootfsEnv = getEnvKey(rootfs);
    const rootfsBin = `${rootfsEnv}/bin`;
    const rootfsEtc = `${rootfsEnv}/etc`;
    const rootfsLib = `${rootfsEnv}/lib`;
    const rootfsLib64 = `${rootfsEnv}/lib64`;
    const rootfsSbin = `${rootfsEnv}/sbin`;
    const rootfsUsr = `${rootfsEnv}/usr`;

    const rootfsArgs = [
      "--ro-bind",
      rootfsBin,
      "/bin",
      "--ro-bind",
      rootfsEtc,
      "/etc",
      "--ro-bind",
      rootfsLib,
      "/lib",
      "--ro-bind-try",
      rootfsLib64,
      "/lib64",
      "--ro-bind",
      rootfsSbin,
      "/sbin",
      "--ro-bind",
      rootfsUsr,
      "/usr",
    ];

    stepArguments.push(...rootfsArgs);
    stepArtifacts.push(rootfs);
  }

  // Setup artifact arguments
  for (const artifact of artifacts) {
    stepArtifacts.push(artifact);
  }

  for (const artifact of stepArtifacts) {
    stepArguments.push("--ro-bind");
    stepArguments.push(getEnvKey(artifact));
    stepArguments.push(getEnvKey(artifact));
    stepArguments.push("--setenv");
    stepArguments.push(getEnvKey(artifact).replace("$", ""));
    stepArguments.push(getEnvKey(artifact));
  }

  // Setup environment arguments
  const stepPathBins = stepArtifacts
    .map((a) => `${getEnvKey(a)}/bin`)
    .join(":");

  let stepPath = `${stepPathBins}:/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin`;

  const pathEnv = environments.find((x) => x.startsWith("PATH="));
  if (pathEnv) {
    const pathValue = pathEnv.split("=").slice(1).join("=");
    if (pathValue) {
      stepPath = `${pathValue}:${stepPath}`;
    }
  }

  stepArguments.push("--setenv");
  stepArguments.push("PATH");
  stepArguments.push(stepPath);

  for (const env of environments) {
    const eqIdx = env.indexOf("=");
    const key = eqIdx !== -1 ? env.substring(0, eqIdx) : env;
    const value = eqIdx !== -1 ? env.substring(eqIdx + 1) : "";

    if (key.startsWith("PATH")) {
      continue;
    }

    stepArguments.push("--setenv");
    stepArguments.push(key);
    stepArguments.push(value);
  }

  // Setup arguments
  for (const argument of arguments_) {
    stepArguments.push(argument);
  }

  // Setup script
  const stepScript = `#!/bin/bash\nset -euo pipefail\n\n${script}\n`;

  // Deduplicate secrets by name
  const uniqueSecrets: ArtifactStepSecret[] = [];
  for (const secret of secrets) {
    if (!uniqueSecrets.some((s) => s.name === secret.name)) {
      uniqueSecrets.push(secret);
    }
  }

  // Setup step
  return {
    entrypoint: "bwrap",
    script: stepScript,
    secrets: uniqueSecrets,
    arguments: stepArguments,
    artifacts: stepArtifacts,
    environments: ["PATH=/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin"],
  };
}

/**
 * Creates a shell step. Matches Rust sdk/rust/src/artifact/step.rs shell().
 *
 * Dispatches to bash() on Darwin, bwrap() on Linux with linux-vorpal rootfs.
 */
export async function shell(
  context: ConfigContext,
  artifacts: string[],
  environments: string[],
  script: string,
  secrets: ArtifactStepSecret[],
): Promise<ArtifactStepMsg> {
  const stepSystem = context.getSystem();

  if (
    stepSystem === ArtifactSystem.AARCH64_DARWIN ||
    stepSystem === ArtifactSystem.X8664_DARWIN
  ) {
    return bash(artifacts, environments, secrets, script);
  }

  if (
    stepSystem === ArtifactSystem.AARCH64_LINUX ||
    stepSystem === ArtifactSystem.X8664_LINUX
  ) {
    const linuxVorpal = await context.fetchArtifactAlias(
      "library/linux-vorpal:latest",
    );

    return bwrap([], artifacts, environments, linuxVorpal, secrets, script);
  }

  throw new Error(`unsupported system: ${stepSystem}`);
}

/**
 * Creates a docker step. Matches Rust sdk/rust/src/artifact/step.rs docker().
 */
export function docker(
  arguments_: string[],
  artifacts: string[],
): ArtifactStepMsg {
  return {
    entrypoint: "docker",
    script: undefined,
    secrets: [],
    arguments: arguments_,
    artifacts: artifacts,
    environments: ["PATH=/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin"],
  };
}
